File: src/src_core/fmg/core/img/AnimatedInnerController.java
package fmg.core.img;

import java.beans.PropertyChangeEvent;
import java.beans.PropertyChangeListener;
import java.util.HashMap;
import java.util.Map;

import fmg.common.Logger;
import fmg.common.geom.SizeDouble;
import fmg.common.ui.UiInvoker;

/**
 * MVC controller. Base animation controller.
 * @param <TImage> platform specific view/image/picture or other display context/canvas/window/panel
 * @param <TImageView> MVC view
 * @param <TImageModel> MVC model
 */
public final class AnimatedInnerController<TImage,
                                           TImageView  extends IImageView<TImage, TImageModel>,
                                           TImageModel extends IAnimatedModel>
            implements IAnimatedController<TImage, TImageView, TImageModel>
{

    private final TImageModel _model;
    private Map<Class<? extends IModelTransformer>, IModelTransformer> _transformers = new HashMap<>();
    private boolean _animationWasUsed = false;
    private final PropertyChangeListener onModelPropertyChangedListener = this::onModelPropertyChanged;

    public AnimatedInnerController(TImageModel model) {
        _model = model;
        model.addListener(onModelPropertyChangedListener);
    }


    // #region: begin unusable code
    @Override
    public TImageModel getModel()                               { throw new UnsupportedOperationException(); }
    @Override
    public TImage getImage()                                    { throw new UnsupportedOperationException(); }
    @Override
    public SizeDouble getSize()                                 { throw new UnsupportedOperationException(); }
    @Override
    public void addListener(PropertyChangeListener listener)    { throw new UnsupportedOperationException(); }
    @Override
    public void removeListener(PropertyChangeListener listener) { throw new UnsupportedOperationException(); }
    @Override
    public void useRotateTransforming(boolean enable)           { throw new UnsupportedOperationException(); }
    @Override
    public void usePolarLightFgTransforming(boolean enable)     { throw new UnsupportedOperationException(); }
    // #region: end unusable code


    @Override
    public void addModelTransformer(IModelTransformer transformer) {
        if (!_transformers.keySet().contains(transformer.getClass()))
            _transformers.put(transformer.getClass(), transformer);
    }
    @Override
    public void removeModelTransformer(Class<? extends IModelTransformer> transformerClass) {
        if (_transformers.keySet().contains(transformerClass))
            _transformers.remove(transformerClass);
    }

    private void onModelPropertyChanged(PropertyChangeEvent ev) {
        switch (ev.getPropertyName()) {
        case IAnimatedModel.PROPERTY_ANIMATED:
            _animationWasUsed = true;
            if ((Boolean)ev.getNewValue()) {
                TImageModel model = _model;
                UiInvoker.ANIMATOR.get().subscribe(this, timeFromStartSubscribe -> {
                    long mod = timeFromStartSubscribe % model.getAnimatePeriod();
                    long frame = mod * model.getTotalFrames() / model.getAnimatePeriod();
                    //Logger.info("ANIMATOR : " + getClass().getSimpleName() + ": "+ timeFromStartSubscribe);
                    model.setCurrentFrame((int)frame);
                });
            } else {
                UiInvoker.ANIMATOR.get().pause(this);
            }
            break;
        case IAnimatedModel.PROPERTY_CURRENT_FRAME:
            if (_transformers.size() == 0)
                Logger.error("No any transformer! " + getClass().getSimpleName()); // зачем работать анимации если нет трансформеров модели
            _transformers.forEach((k,v) -> v.execute(_model));
            break;
        }
    }

    @Override
    public void close() {
        _model.removeListener(onModelPropertyChangedListener);
        if (_animationWasUsed) // do not call UiInvoker.ANIMATOR if it is not already used
            UiInvoker.ANIMATOR.get().unsubscribe(this);
        _transformers.clear();
    }

}

File: src/src_core/fmg/core/img/AnimatedInnerController.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using fmg.common.geom;
using fmg.common.notifier;
using fmg.common.ui;

namespace fmg.core.img {

    /// <summary> MVC controller. Base animation controller. </summary>
    /// <typeparam name="TImage">platform specific view/image/picture or other display context/canvas/window/panel</typeparam>
    /// <typeparam name="TImageView">MVC view</typeparam>
    /// <typeparam name="TImageModel">MVC model</typeparam>
    public sealed class AnimatedInnerController<TImage, TImageView, TImageModel>
                          : IAnimatedController<TImage, TImageView, TImageModel>
        where TImage : class
        where TImageView : IImageView<TImage, TImageModel>
        where TImageModel : IAnimatedModel
    {
        private readonly TImageModel _model;
        private IDictionary<Type, IModelTransformer> _transformers = new Dictionary<Type, IModelTransformer>();
#pragma warning disable CS0067 // warning CS0067: The event is never used
        public event PropertyChangedEventHandler PropertyChanged; // TODO unusable
#pragma warning restore CS0067
        private bool _animationWasUsed = false;

        public AnimatedInnerController(TImageModel model) {
            _model = model;
            _model.PropertyChanged += OnPropertyModelChanged;
        }


        #region: begin unusable code
        public TImageModel Model { get { throw new NotImplementedException(); } }
        public TImage      Image { get { throw new NotImplementedException(); } }
        public SizeDouble  Size  { get { throw new NotImplementedException(); } }
        public void UseRotateTransforming(bool enable)       { throw new NotImplementedException(); }
        public void UsePolarLightFgTransforming(bool enable) { throw new NotImplementedException(); }
        #endregion: end unusable code


        public void RemoveModelTransformer(Type /* extends IModelTransformer */ transformerClass) {
            if (_transformers.ContainsKey(transformerClass))
                _transformers.Remove(transformerClass);
        }
        public void AddModelTransformer(IModelTransformer transformer) {
            if (!_transformers.ContainsKey(transformer.GetType()))
                _transformers.Add(transformer.GetType(), transformer);
        }

        private void OnPropertyModelChanged(object sender, PropertyChangedEventArgs ev) {
            switch (ev.PropertyName) {
            case nameof(IAnimatedModel.Animated):
                _animationWasUsed = true;
                if ((ev as PropertyChangedExEventArgs<bool>).NewValue) {
                    TImageModel model = _model;
                    UiInvoker.Animator().Subscribe(this, timeFromStartSubscribe => {
                        var mod = timeFromStartSubscribe.TotalMilliseconds % model.AnimatePeriod;
                        var frame = mod * model.TotalFrames / model.AnimatePeriod;
                        model.CurrentFrame = (int)frame;
                        //System.Diagnostics.Debug.WriteLine("ANIMATOR : CurrentFrame" + frame + "/" + model.TotalFrames);
                    });
                } else {
                    UiInvoker.Animator().Pause(this);
                }
                break;
            case nameof(IAnimatedModel.CurrentFrame):
                if (!_transformers.Any())
                    System.Diagnostics.Debug.WriteLine("No any transformer! " + GetType().Name); // зачем работать анимации если нет трансформеров модели
                foreach (var item in _transformers)
                    item.Value.Execute(_model);
                break;
            }
        }

        public void Dispose() {
            _model.PropertyChanged -= OnPropertyModelChanged;
            if (_animationWasUsed) // do not call UiInvoker.Animator if it is not already used
                UiInvoker.Animator().Unsubscribe(this);
            _transformers.Clear();
            GC.SuppressFinalize(this);
        }

    }

}
